File: leetcodes/recursion/Q78.cpp
#include <vector>
using namespace std;

class Solution
{
public:
    vector<int> temp;
    vector<vector<int>> answer;
    vector<vector<int>> subsets(vector<int> &nums)
    {
        helper(0, nums);
        return answer;
    }

    void helper(int current, vector<int> &nums)
    {
        if (current == nums.size())
        {
            answer.push_back(temp);
            return;
        }
        temp.push_back(nums[current]);
        helper(current + 1, nums);
        temp.pop_back();
        helper(current + 1, nums);
    }
};

File: leetcodes/backTrace/Q78.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        self.ans = []
        self.subset = []
        def helper(index):
            if index==len(nums):
                self.ans.append(self.subset.copy())
                return
            # put
            self.subset.append(nums[index])
            helper(index+1)
            self.subset.pop()
            helper(index+1)
            return
        helper(0)
        return self.ans